File: D-MASH/client/backend/route_discovery_v4.py
"""Opaque route discovery proofs with independent delegated discovery keys.

Callers pin certificates via authenticated route descriptors; this module neither
provisions delegation nor grants mailbox/Account authority.
"""
import base64
import hashlib
import json
import re
import secrets
import struct
import time

from nacl.public import Box, PrivateKey, PublicKey
from nacl.signing import VerifyKey

CERT_FIELDS = {'version','route_id','discovery_sign','discovery_box','recipient_box','generation','issued_at','expires_at','signature'}
MAX_LIFETIME = 30 * 86400
MAX_BLOB = 16384


def _hex(value, size=32):
    if not isinstance(value,str) or not re.fullmatch('[0-9a-f]{%d}' % (size*2), value):
        raise ValueError('Invalid discovery encoding')
    return bytes.fromhex(value)


def _integer(value):
    if type(value) is not int or not 0 <= value <= 2**53-1:
        raise ValueError('Invalid discovery integer')
    return struct.pack('>Q',value)


def certificate_transcript(cert):
    if not isinstance(cert,dict) or set(cert)!=CERT_FIELDS or type(cert['version']) is not int or cert['version']!=4:
        raise ValueError('Invalid discovery certificate')
    keys=b''.join(_hex(cert[name]) for name in ('route_id','discovery_sign','discovery_box','recipient_box'))
    dates=b''.join(_integer(cert[name]) for name in ('generation','issued_at','expires_at'))
    if cert['generation']<1 or not 0<cert['expires_at']-cert['issued_at']<=MAX_LIFETIME:
        raise ValueError('Invalid discovery validity')
    return b'D-MASH|DISCOVERY-CERT|V4\0'+keys+dates


def verify_certificate(cert,now=None):
    now=int(time.time()) if now is None else now
    transcript=certificate_transcript(cert)
    if not cert['issued_at']<=now+60 or cert['expires_at']<=now:
        raise ValueError('Discovery certificate expired')
    VerifyKey(_hex(cert['route_id'])).verify(transcript,_hex(cert['signature'],64))
    return transcript


def issue_certificate(route_sign,discovery_sign,discovery_box,recipient_box,*,generation,issued_at,expires_at):
    cert=dict(version=4,route_id=route_sign.verify_key.encode().hex(),discovery_sign=bytes(discovery_sign).hex(),
              discovery_box=bytes(discovery_box).hex(),recipient_box=bytes(recipient_box).hex(),
              generation=generation,issued_at=issued_at,expires_at=expires_at,signature='')
    cert['signature']=route_sign.sign(certificate_transcript(cert)).signature.hex()
    return cert


def seal(public_key,value):
    raw=json.dumps(value,separators=(',',':'),ensure_ascii=True).encode()
    if len(raw)>MAX_BLOB-72:raise ValueError('Discovery payload too large')
    ephemeral=PrivateKey.generate();nonce=secrets.token_bytes(24)
    cipher=Box(ephemeral,PublicKey(public_key)).encrypt(raw,nonce).ciphertext
    return base64.b64encode(bytes(ephemeral.public_key)+nonce+cipher).decode()


def open_box(private_key,value):
    if not isinstance(value,str) or len(value)>((MAX_BLOB+2)//3)*4:raise ValueError('Invalid discovery box')
    raw=base64.b64decode(value,validate=True)
    if not 72<=len(raw)<=MAX_BLOB or base64.b64encode(raw).decode()!=value:raise ValueError('Invalid discovery box')
    plain=Box(PrivateKey(bytes(private_key)),PublicKey(raw[:32])).decrypt(raw[56:],raw[32:56])
    return json.loads(plain)


def _query(query,cert,now):
    if not isinstance(query,dict) or set(query)!={'type','version','route_id','challenge','reply_key','expires_at'}:
        raise ValueError('Invalid discovery query')
    if query['type']!='ROUTE_QUERY' or type(query['version']) is not int or query['version']!=4 or query['route_id']!=cert['route_id']:
        raise ValueError('Invalid discovery context')
    _hex(query['challenge']);_hex(query['reply_key']);_integer(query['expires_at'])
    if not now<query['expires_at']<=min(now+180,cert['expires_at']):raise ValueError('Discovery query expired')
    return query


def _reply_transcript(query,cert):
    digest=hashlib.sha256(certificate_transcript(cert)+_hex(cert['signature'],64)).digest()
    return b'D-MASH|DISCOVERY-REPLY|V4\0'+_hex(query['challenge'])+_hex(query['reply_key'])+_integer(query['expires_at'])+digest


def create_query(cert,*,now=None):
    now=int(time.time()) if now is None else now
    verify_certificate(cert,now);reply=PrivateKey.generate()
    query=dict(type='ROUTE_QUERY',version=4,route_id=cert['route_id'],challenge=secrets.token_hex(32),
               reply_key=bytes(reply.public_key).hex(),expires_at=min(now+180,cert['expires_at']))
    return seal(_hex(cert['discovery_box']),query),dict(query=query,certificate=dict(cert),reply_private=reply)


def answer_query(blob,cert,discovery_sign,discovery_box,*,now=None,with_context=False):
    now=int(time.time()) if now is None else now
    verify_certificate(cert,now)
    if discovery_sign.verify_key.encode().hex()!=cert['discovery_sign'] or bytes(discovery_box.public_key).hex()!=cert['discovery_box']:
        raise ValueError('Discovery key mismatch')
    query=_query(open_box(discovery_box,blob),cert,now)
    response=dict(type='ROUTE_REPLY',version=4,challenge=query['challenge'],expires_at=query['expires_at'],
                  certificate=cert,signature=discovery_sign.sign(_reply_transcript(query,cert)).signature.hex())
    reply=seal(_hex(query['reply_key']),response)
    return (reply,query['expires_at']) if with_context else reply


def verify_reply(blob,state,*,now=None):
    now=int(time.time()) if now is None else now
    cert=state['certificate'];verify_certificate(cert,now);query=_query(state['query'],cert,now)
    response=open_box(state['reply_private'],blob)
    if not isinstance(response,dict) or set(response)!={'type','version','challenge','expires_at','certificate','signature'}:
        raise ValueError('Invalid discovery reply')
    if (response['type']!='ROUTE_REPLY' or type(response['version']) is not int or response['version']!=4
            or response['certificate']!=cert or response['challenge']!=query['challenge']
            or type(response['expires_at']) is not int or response['expires_at']!=query['expires_at']):
        raise ValueError('Discovery reply context mismatch')
    verify_certificate(response['certificate'],now)
    VerifyKey(_hex(cert['discovery_sign'])).verify(_reply_transcript(query,cert),_hex(response['signature'],64))
    return True

File: D-MASH PWA/not_messenger/js/route_discovery_v4.js
'use strict';
(function(global){
 const MAX_BLOB=16384,MAX_LIFETIME=30*86400;
 const text=value=>new TextEncoder().encode(value);
 const join=(...parts)=>{const out=new Uint8Array(parts.reduce((n,p)=>n+p.length,0));let at=0;for(const part of parts){out.set(part,at);at+=part.length;}return out;};
 const hex=value=>Array.from(value,b=>b.toString(16).padStart(2,'0')).join('');
 const unhex=(value,size=32)=>{if(typeof value!=='string'||!(new RegExp('^[0-9a-f]{'+size*2+'}$')).test(value))throw Error('Invalid discovery encoding');return Uint8Array.from(value.match(/../g),b=>parseInt(b,16));};
 const integer=value=>{if(!Number.isSafeInteger(value)||value<0)throw Error('Invalid discovery integer');const out=new Uint8Array(8);new DataView(out.buffer).setBigUint64(0,BigInt(value));return out;};
 const exact=(value,keys)=>value&&typeof value==='object'&&!Array.isArray(value)&&Object.keys(value).sort().join(',')===keys;
 function certificateTranscript(cert){
  if(!exact(cert,'discovery_box,discovery_sign,expires_at,generation,issued_at,recipient_box,route_id,signature,version')||cert.version!==4)throw Error('Invalid discovery certificate');
  const keys=['route_id','discovery_sign','discovery_box','recipient_box'].map(name=>unhex(cert[name]));
  const dates=['generation','issued_at','expires_at'].map(name=>integer(cert[name]));
  if(cert.generation<1||cert.expires_at<=cert.issued_at||cert.expires_at-cert.issued_at>MAX_LIFETIME)throw Error('Invalid discovery validity');
  return join(text('D-MASH|DISCOVERY-CERT|V4\0'),...keys,...dates);
 }
 function verifyCertificate(cert,now=Math.floor(Date.now()/1000)){
  const transcript=certificateTranscript(cert);
  if(cert.issued_at>now+60||cert.expires_at<=now)throw Error('Discovery certificate expired');
  if(!global.nacl.sign.detached.verify(transcript,unhex(cert.signature,64),unhex(cert.route_id)))throw Error('Invalid discovery certificate signature');
  return transcript;
 }
 function issueCertificate(routeSign,discoverySign,discoveryBox,recipientBox,{generation,issuedAt,expiresAt}){
  const cert={version:4,route_id:hex(routeSign.publicKey),discovery_sign:hex(discoverySign),discovery_box:hex(discoveryBox),recipient_box:hex(recipientBox),generation,issued_at:issuedAt,expires_at:expiresAt,signature:''};
  cert.signature=hex(global.nacl.sign.detached(certificateTranscript(cert),routeSign.secretKey));return cert;
 }
 const b64=value=>btoa(String.fromCharCode(...value));
 function seal(publicKey,value){
  const raw=text(JSON.stringify(value));if(raw.length>MAX_BLOB-72)throw Error('Discovery payload too large');
  const ephemeral=global.nacl.box.keyPair(),nonce=global.nacl.randomBytes(24);
  try{return b64(join(ephemeral.publicKey,nonce,global.nacl.box(raw,nonce,publicKey,ephemeral.secretKey)));}
  finally{ephemeral.secretKey.fill(0);}
 }
 function openBox(privateKey,value){
  if(typeof value!=='string'||value.length>Math.ceil(MAX_BLOB/3)*4)throw Error('Invalid discovery box');
  const raw=Uint8Array.from(atob(value),c=>c.charCodeAt(0));
  if(raw.length<72||raw.length>MAX_BLOB||b64(raw)!==value)throw Error('Invalid discovery box');
  const plain=global.nacl.box.open(raw.subarray(56),raw.subarray(32,56),raw.subarray(0,32),privateKey);
  if(!plain)throw Error('Discovery authentication failed');
  try{return JSON.parse(new TextDecoder('utf-8',{fatal:true}).decode(plain));}finally{plain.fill(0);}
 }
 function checkQuery(query,cert,now){
  if(!exact(query,'challenge,expires_at,reply_key,route_id,type,version')||query.type!=='ROUTE_QUERY'||query.version!==4||query.route_id!==cert.route_id)throw Error('Invalid discovery context');
  unhex(query.challenge);unhex(query.reply_key);integer(query.expires_at);
  if(query.expires_at<=now||query.expires_at>Math.min(now+180,cert.expires_at))throw Error('Discovery query expired');
  return query;
 }
 async function replyTranscript(query,cert){
  const digest=new Uint8Array(await crypto.subtle.digest('SHA-256',join(certificateTranscript(cert),unhex(cert.signature,64))));
  return join(text('D-MASH|DISCOVERY-REPLY|V4\0'),unhex(query.challenge),unhex(query.reply_key),integer(query.expires_at),digest);
 }
 function createQuery(cert,{now=Math.floor(Date.now()/1000)}={}){
  verifyCertificate(cert,now);cert=Object.freeze({...cert});
  const reply=global.nacl.box.keyPair();
  try{
   const query=Object.freeze({type:'ROUTE_QUERY',version:4,route_id:cert.route_id,challenge:hex(global.nacl.randomBytes(32)),reply_key:hex(reply.publicKey),expires_at:Math.min(now+180,cert.expires_at)});
   return {blob:seal(unhex(cert.discovery_box),query),state:{query,certificate:cert,replyPrivate:reply.secretKey}};
  }catch(error){reply.secretKey.fill(0);throw error;}
 }
 async function answerQuery(blob,cert,discoverySign,discoveryBox,options={}){
  const clock=options.clock||(()=>options.now??Math.floor(Date.now()/1000)),now=clock();
  verifyCertificate(cert,now);cert={...cert};
  if(hex(discoverySign.publicKey)!==cert.discovery_sign||hex(discoveryBox.publicKey)!==cert.discovery_box)throw Error('Discovery key mismatch');
  const query=checkQuery(openBox(discoveryBox.secretKey,blob),cert,now);
  const signature=hex(global.nacl.sign.detached(await replyTranscript(query,cert),discoverySign.secretKey));
  checkQuery(query,cert,clock());
  const reply=seal(unhex(query.reply_key),{type:'ROUTE_REPLY',version:4,challenge:query.challenge,expires_at:query.expires_at,certificate:cert,signature});
  return options.withContext?{reply,expiresAt:query.expires_at}:reply;
 }
 async function verifyReply(blob,state,options={}){
  const clock=options.clock||(()=>options.now??Math.floor(Date.now()/1000)),now=clock();
  const cert=state.certificate;verifyCertificate(cert,now);const query=checkQuery(state.query,cert,now);
  const response=openBox(state.replyPrivate,blob);
  if(!exact(response,'certificate,challenge,expires_at,signature,type,version')||response.type!=='ROUTE_REPLY'||response.version!==4||response.challenge!==query.challenge||response.expires_at!==query.expires_at)throw Error('Discovery reply context mismatch');
  verifyCertificate(response.certificate,now);
  if(Object.keys(cert).some(key=>cert[key]!==response.certificate[key]))throw Error('Discovery certificate mismatch');
  if(!global.nacl.sign.detached.verify(await replyTranscript(query,cert),unhex(response.signature,64),unhex(cert.discovery_sign)))throw Error('Invalid discovery reply signature');
  checkQuery(query,cert,clock());
  return true;
 }
 const api=Object.freeze({certificateTranscript,verifyCertificate,issueCertificate,seal,openBox,createQuery,answerQuery,verifyReply});
 global.DmashRouteDiscoveryV4=api;
 if(typeof module!=='undefined')module.exports=api;
})(typeof window!=='undefined'?window:globalThis);
